go-controller/pkg/controller-manager/controllerManager.go

```diff
@@ -0,0 +1,274 @@
+		return err
+	}
+
+	cm.configureMetrics()
+
+	cm.NewDefaultController()
+	return nil
+}
+
+func (cm *NetworkControllerManager) configureSCTPSupport() error {
+	hasSCTPSupport, err := util.DetectSCTPSupport()
+	if err != nil {
+		return err
+	}
+
+	if !hasSCTPSupport {
+		klog.Warningf("SCTP unsupported by this version of OVN. Kubernetes service creation with SCTP will not work ")
+	} else {
+		klog.Info("SCTP support detected in OVN")
+	}
+	cm.SCTPSupport = hasSCTPSupport
+	return nil
+}
+
+// enableOVNLogicalDataPathGroups sets an OVN flag to enable logical datapath
+// groups on OVN 20.12 and later. The option is ignored if OVN doesn't
+// understand it. Logical datapath groups reduce the size of the southbound
+// database in large clusters. ovn-controllers should be upgraded to a version
+// that supports them before the option is turned on by the master.
+func (cm *NetworkControllerManager) enableOVNLogicalDataPathGroups() error {
+	nbGlobal := nbdb.NBGlobal{
+		Options: map[string]string{"use_logical_dp_groups": "true"},
+	}
+	if err := libovsdbops.UpdateNBGlobalSetOptions(cm.nbClient, &nbGlobal); err != nil {
+		return fmt.Errorf("failed to set NB global option to enable logical datapath groups: %v", err)
+	}
+	return nil
+}
+
+func (cm *NetworkControllerManager) configureMetrics() {
+	metrics.RunTimestamp(cm.stopChan, cm.sbClient, cm.nbClient)
+	metrics.MonitorIPSec(cm.nbClient)
+	if config.Metrics.EnableConfigDuration {
+		// with k=10,
+		//  for a cluster with 10 nodes, measurement of 1 in every 100 requests
+		//  for a cluster with 100 nodes, measurement of 1 in every 1000 requests
+		metrics.GetConfigDurationRecorder().Run(cm.nbClient, cm.kube, 10, time.Second*5, cm.stopChan)
+	}
+	cm.podRecorder.Run(cm.sbClient, cm.stopChan)
+}
+
+// NewDefaultController creates and returns the controller for default network
+func (cm *NetworkControllerManager) NewDefaultController() {
+	cc := ovn.NewControllerConnection(cm.client, cm.kube, cm.watchFactory, cm.recorder, cm.nbClient,
+		cm.sbClient, cm.podRecorder, cm.SCTPSupport)
+	defaultController := ovn.NewDefaultController(cc)
+	cm.ovnControllers[ovntypes.DefaultNetworkName] = defaultController
+}
+
+// Run starts to handle all the secondary net-attach-def and creates and manages all the secondary controllers
+func (cm *NetworkControllerManager) Run(ctx context.Context) error {
+	if err := cm.watchFactory.Start(); err != nil {
+		return err
+	}
+
+	defaultController, ok := cm.ovnControllers[ovntypes.DefaultNetworkName]
+	if ok {
+		return defaultController.Start(ctx)
+	}
+
+	// start the net-attach-def controller which handles net-attach-def events and
+	// creates/deletes/starts secondary controllers,
+	return nil
+}
+
+// Stop gracefully stops all managed controllers
+func (cm *NetworkControllerManager) Stop() {
+	// stop the net-attach-def controller
+	// and for each Controller of secondary network, call oc.Stop()
+	for _, oc := range cm.ovnControllers {
+		oc.Stop()
+	}
+}
```

Comment:
It still feels redundant to have Stop() method and provide a stopChan to NetworkControllerManager. NetworkControllerManager could have its own private stopChan that is closed in this Stop() method.

Reply:
But please note as well that having a single stopchan and WG for all controllers can make terminating the process slower. With one Stop() per controller with a wait inside, controllers are closed in sequence, and in this PR is not really apparent why we need to do it in sequence vs closing all at the same time.

Reply:
secondary controllers must have their own stopChan, as they may need to be stopped when all net-attach-def of that network are deleted.

Reply:
Make sense. But how does that apply to the NetworkControllerManager.

Reply:
looks like the stopChan is private to NetworkControllerManager and is just used for metrics. It doesn't need to be a property of the NetworkControllerManager. You can just remove it from NewNetworkControllerManager, and pass it as part of cm.Start(ctx.Context, cancel, stopChan)->cm.Init(stopChan) -> cm.configureMetrics(stopChan)

Reply:
My original comment though is that we don't really need to pass a stopChan to NetworkControllerManager.
NetworkControllerManager can create it's own stopChan internally and close it when NetworkControllerManager.Stop() is called.

Currently NetworkControllerManager is provided with 3 different possible ways of being told to stop:
1. A cancel-able context
2. A stopChan
3. A Stop() method

We should only ever need 1. We had 2 (context & stopChan). We added a 3rd (Stop() method) in this PR. I am suggesting a way to keep it at 2 although it would not be too complicated to just keep the Stop() method and spawn the cancel-able context and stopChan internally for whichever downstream APIs require it, and cancel/close them when Stop() is called.